### PR TITLE
Replace Game Server with WebViews

### PR DESCRIPTION
**Pull request by CareyJWilliams**
Replaces the game server components and browser game running with a simpler system that uses the webview API built into VS Code.

The notable/additions changes are:
- ChoiceScript Compiler
- The Game Web View
- An Abstraction layer over the VSCode workspace namespace (VS Code Workspace Provider) to help writing tests in the client (compiler tests in this case)*
- Renamed "Open Game" to "Run Game"
- Added an action modal that triggers if you try to run a game when one is already running

*This one in particular does cause a bit of churn, but I think it's a necessary step to enable better tested clients.

Otherwise all the behaviour should remain exactly the same, e.g. the way errors are annotated etc.

**Screenshots**

Screenshot 2024-04-27 at 17 28 53
Screenshot 2024-04-27 at 17 29 00

### CODE REVIEW

**Comment by sgranade** [comment]
I did a clean pull, `npm install`, and Build/Run task in VS Code. When I do, and then run the "Open Game" command, I get a page that says it's loading but never actually loads.

image

Any guesses about what might be happening? I can send you my game files if you'd like to try it out.

**Comment by CareyJWilliams** [comment]
> I did a clean pull, `npm install`, and Build/Run task in VS Code. When I do, and then run the "Open Game" command, I get a page that says it's loading but never actually loads.
> 
> image
> 
> Any guesses about what might be happening? I can send you my game files if you'd like to try it out.

If you open the web view developer tools via the command "Open Webview Developer Tools", you'll probably find you've got some informative error in the console. I can't guess what that'd be right now. Out of town until tomorrow evening, when I'll double check on my side. If you can't find anything via the above method, please do pass your files over and I'll be sure to take a look.

**Comment by sgranade** [comment]
Aha, looks like there's a problem loading the initial scene file. I'm going to go out on a limb and guess it's an issue with Windows files. I'll dig into it tomorrow and see what I can find.

File: client/src/common/choicescript-compiler.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	constructor(
+		private readonly workspaceProvider: IWorkspaceProvider
+	) {}
+
+	public async compile(scenes: Uri[]): Promise<CompiledChoiceScriptGame>{
+		const allScenes: AllScenesResult = {};
+		let title: string;
+	
+		for (const s of scenes) {
+			const sceneName = s.fsPath.substring(s.fsPath.lastIndexOf('/') + 1, (s.fsPath.length - ".txt".length));
```
Comment (sgranade) [inline client/src/common/choicescript-compiler.ts:24]: I figured out: this line is why it's failing on Windows systems. The URIs getting passed on Windows look like:

` c:\\Users\\username\\Documents\\game\\startup`

The code expects them to be Posix-style with forward slashes. We'll need to accommodate both Posix and Windows-style URIs.

**Comment by sgranade** [comment]
Progress! I've now got my game running in a webview and it looks great. Error annotations aren't coming back through for me, though. I forced an error which showed up in the console but didn't trigger an annotation.

image

File: client/src/common/choicescript-compiler.ts (hunk 1 above)
Comment (CareyJWilliams) [reply client/src/common/choicescript-compiler.ts]: How sloppy of me!

Comment (CareyJWilliams) [reply client/src/common/choicescript-compiler.ts]: Far too focused on making it work for web and desktop that I wasn't considering different OSes probably.

**Comment by CareyJWilliams** [comment]
> Progress! I've now got my game running in a webview and it looks great. Error annotations aren't coming back through for me, though. I forced an error which showed up in the console but didn't trigger an annotation.
> 
> image

I'll double check this tonight!

**Comment by CareyJWilliams** [comment]
@sgranade 

It's still working for me. I'm not sure what could be wrong there.

Kapture 2024-05-13 at 18 51 57

**Comment by CareyJWilliams** [comment]
Unfortunately I don't have easy access to a Windows machine at present.

**Comment by sgranade** [comment]
I'm finally back on this today! While I'm chasing down why error annotations don't show up for me, can you investigate what it'd take to provide support for images? Right now they're being mapped to a webview URI that is then blocked.

image

**Comment by CareyJWilliams** [comment]
> I'm finally back on this today! While I'm chasing down why error annotations don't show up for me, can you investigate what it'd take to provide support for images? Right now they're being mapped to a webview URI that is then blocked.
> 
> image

Is it just when images are in use that things are broken? Should be an easy fix...

**Comment by sgranade** [comment]
The error I'm seeing is from the trying to load an image via `*text_image`. I'm guessing we need to do some alterations to image URLs in `getWebviewContent()`, potentially by pulling in the local image URIs from the language server. The old setup supported image files either local to the scene files or one directory up, since Dashingdon wants the former while a local CS setup expects the latter.

In the meantime, I finally figured out why annotations didn't work for me: at some point I broke my scripts that minified CS in my local repo. I fixed that and now everything's good on my end.

I'm going to update the packages in prep to merge this change in and move directly to releasing a new version of the extension.

**Comment by CareyJWilliams** [comment]
@sgranade I've pushed a patch to support images and sounds (in theory, although they won't work due to another [problem](https://github.com/microsoft/vscode/issues/197937)).

I'll be honest, I'm not 100% happy with this approach. But this turned out to be a surprisingly complicated problem to solve.
Ultimately all resources loaded by a webview need their URI to be fed through asWebviewUri. This is fine for discrete and static assets but a real pain for anything added dynamically by the user.

I could think of three valid approaches to solving this problem:

1. Reverting to having a server, albeit just for media assets -- but that's still difficult because you'd need to somehow inject the server's address into user code (e.g. via modification to ChoiceScript) 
2. Convert all images to base64 or data URIs -- similar problem to above in that we don't definitively know the full image paths at "compile" time, e.g. image paths may be presented as variables so the full path often isn't apparent until runtime.
3. Use injected code to watch for relevant assets and send a request back to the extension to swap out the URIs -- this is the approach I've gone for. 

It's not pretty, but I've spent a lot of time trying to come up with something half decent and I do think this approach is the lesser of the three evils. I'm happy to hear any other ideas you might have though.

**Pull request merged**